portfolio/src/main/webapp/script.js

```diff
@@ -203,4 +201,252 @@ function getFilter() {
+    const addButton = document.createElement('button');
+    addButton.innerText = "Add";
+
+    addButton.onclick = () => {
+        sendMarker(lat, lng);
+        editMarker.setMap(null);
+    };
+    return addButton;
+}
+
+/** Sends marker to servlet */
+function sendMarker(lat, lng) {
+    const url = "/markers?lat=" + lat + "&lng=" + lng;
+    const request = new Request(url, {method: 'POST'});
+    
+    fetch(request).then((response) => response.json()).then((marker) => {
+        displayMarker(marker.lat, marker.lng, marker.userId, marker.id);
+    });
+}
+
+/** Helper function that displays markers */
+function displayMarker(lat, lng, userId, id) {
+    const marker = new google.maps.Marker({
+        position: {lat: lat, lng: lng},
+        map: map,
+        userId: userId,
+        id: id});
+    markersMap.set(id, marker);
+
+    var infoWindow = new google.maps.InfoWindow;
+    infoWindow.addListener('closeclick', () => {
+        infoWindowsOpened--;
+    })
+
+    var geocoder = new google.maps.Geocoder;
+    marker.addListener('click', () => {
+        openInfoWindow(map, marker, infoWindow, geocoder);
+    });
+}
+
+/** Builds and Opens Info Window */
+function openInfoWindow(map, marker, infoWindow, geocoder) {
+    const latlng = marker.get('position');
+    const textNode = document.createElement('div');
+    const windowNode = document.createElement('div');
+
+    // Retrieves general location address and updates textNode
+    geocoder.geocode({'location': latlng}, function(results, status) {
+        if (status == 'OK') {
+            if (results[0]) {
+                var result = results[0]
+                var city = "";
+                var state = "";
+                var country = "";
+                for (var i = 0; i < result.address_components.length; i++) {
+                    var component = result.address_components[i];
+                    if ((city == "") && (component.types.indexOf("political") >= 0 )) {
+                        city = component.long_name + ", ";
+                    }
+                    if ((state == "") && (component.types.indexOf("administrative_area_level_1") >= 0 )) {
+                        state = component.short_name + ", ";
+                    }
+                    if ((country == "") && (component.types.indexOf("country") >= 0 )) {
+                        country = component.long_name;
+                    }
+                }
+                textNode.innerText = city + state + country;
+            }
+        }
+    });
+
+    windowNode.appendChild(textNode);
+
+    // gives users who created the marker the option to delete it
+    if (marker.get('userId') == document.body.dataset.userId) {
+        const deleteButton = document.createElement('button');
+        deleteButton.innerText = "Delete";
+        deleteButton.onclick = () => {
+            markersMap.get(marker.get('id')).setMap(null);
+            markersMap.delete(marker.get('id'));
```

Comment:
Maybe `deleteMarker` ought to handle removing the marker from the map, as well as from the database.

Reply:
That would work better, I'll move things around to achieve all of this within deleteMarker(). Thanks!